File: Erlang/csvparser.erl
%% Copyright 2012 agwlvssainokuni
%%
%% Licensed under the Apache License, Version 2.0 (the "License");
%% you may not use this file except in compliance with the License.
%% You may obtain a copy of the License at
%%
%%     http://www.apache.org/licenses/LICENSE-2.0
%%
%% Unless required by applicable law or agreed to in writing, software
%% distributed under the License is distributed on an "AS IS" BASIS,
%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%% See the License for the specific language governing permissions and
%% limitations under the License.

-module(csvparser).
-export([read/2, file/1, string/1]).


%% @spec read(Fun, Arg) => {ok, [Field, Field, ...}, NextArg}
%%                         {eof, NextArg}
read(Fun, Arg) ->
	case read(Fun, Arg, [], [], state_RECORD_BEGIN) of
		{ok, [], Arg2} -> {eof, Arg2};
		Else -> Else
	end.

read(Fun, Arg, F, R, state_RECORD_BEGIN) ->
	try Fun(Arg) of
		{ok, ",", Arg2} -> read(Fun, Arg2, F, R, $,,
				action_FLUSH,
				state_FIELD_BEGIN);
		{ok, "\"", Arg2} -> read(Fun, Arg2, F, R, $",
				action_NONE,
				state_ESCAPED);
		{ok, "\r", Arg2} -> read(Fun, Arg2, F, R, $\r,
				action_FLUSH,
				state_CR);
		{ok, "\n", Arg2} -> read(Fun, Arg2, F, R, $\n,
				action_FLUSH,
				state_RECORD_END);
		{ok, [C], Arg2} -> read(Fun, Arg2, F, R, C,
				action_APPEND,
				state_NONESCAPED);
		{eof, Arg2} -> read(Fun, Arg2, F, R, eof,
				action_NONE,
				state_RECORD_END)
	catch
		_:Why -> {ioerror, Why}
	end;

read(Fun, Arg, F, R, state_FIELD_BEGIN) ->
	try Fun(Arg) of
		{ok, ",", Arg2} -> read(Fun, Arg2, F, R, $,,
				action_FLUSH,
				state_FIELD_BEGIN);
		{ok, "\"", Arg2} -> read(Fun, Arg2, F, R, $",
				action_NONE,
				state_ESCAPED);
		{ok, "\r", Arg2} -> read(Fun, Arg2, F, R, $\r,
				action_FLUSH,
				state_CR);
		{ok, "\n", Arg2} -> read(Fun, Arg2, F, R, $\n,
				action_FLUSH,
				state_RECORD_END);
		{ok, [C], Arg2} -> read(Fun, Arg2, F, R, C,
				action_APPEND,
				state_NONESCAPED);
		{eof, Arg2} -> read(Fun, Arg2, F, R, eof,
				action_FLUSH,
				state_RECORD_END)
	catch
		_:Why -> {ioerror, Why}
	end;

read(Fun, Arg, F, R, state_NONESCAPED) ->
	try Fun(Arg) of
		{ok, ",", Arg2} -> read(Fun, Arg2, F, R, $,,
				action_FLUSH,
				state_FIELD_BEGIN);
		{ok, "\"", Arg2} -> read(Fun, Arg2, F, R, $",
				action_APPEND,
				state_NONESCAPED);
		{ok, "\r", Arg2} -> read(Fun, Arg2, F, R, $\r,
				action_FLUSH,
				state_CR);
		{ok, "\n", Arg2} -> read(Fun, Arg2, F, R, $\n,
				action_FLUSH,
				state_RECORD_END);
		{ok, [C], Arg2} -> read(Fun, Arg2, F, R, C,
				action_APPEND,
				state_NONESCAPED);
		{eof, Arg2} -> read(Fun, Arg2, F, R, eof,
				action_FLUSH,
				state_RECORD_END)
	catch
		_:Why -> {ioerror, Why}
	end;

read(Fun, Arg, F, R, state_ESCAPED) ->
	try Fun(Arg) of
		{ok, ",", Arg2} -> read(Fun, Arg2, F, R, $,,
				action_APPEND,
				state_ESCAPED);
		{ok, "\"", Arg2} -> read(Fun, Arg2, F, R, $",
				action_NONE,
				state_DQUOTE);
		{ok, "\r", Arg2} -> read(Fun, Arg2, F, R, $\r,
				action_APPEND,
				state_ESCAPED);
		{ok, "\n", Arg2} -> read(Fun, Arg2, F, R, $\n,
				action_APPEND,
				state_ESCAPED);
		{ok, [C], Arg2} -> read(Fun, Arg2, F, R, C,
				action_APPEND,
				state_ESCAPED);
		{eof, Arg2} -> read(Fun, Arg2, F, R, eof,
				action_ERROR,
				state_ERROR)
	catch
		_:Why -> {ioerror, Why}
	end;

read(Fun, Arg, F, R, state_DQUOTE) ->
	try Fun(Arg) of
		{ok, ",", Arg2} -> read(Fun, Arg2, F, R, $,,
				action_FLUSH,
				state_FIELD_BEGIN);
		{ok, "\"", Arg2} -> read(Fun, Arg2, F, R, $",
				action_APPEND,
				state_ESCAPED);
		{ok, "\r", Arg2} -> read(Fun, Arg2, F, R, $\r,
				action_FLUSH,
				state_CR);
		{ok, "\n", Arg2} -> read(Fun, Arg2, F, R, $\n,
				action_FLUSH,
				state_RECORD_END);
		{ok, [C], Arg2} -> read(Fun, Arg2, F, R, C,
				action_ERROR,
				state_ERROR);
		{eof, Arg2} -> read(Fun, Arg2, F, R, eof,
				action_FLUSH,
				state_RECORD_END)
	catch
		_:Why -> {ioerror, Why}
	end;

read(Fun, Arg, F, R, state_CR) ->
	try Fun(Arg) of
		{ok, ",", Arg2} -> read(Fun, Arg2, F, R, $,,
				action_ERROR,
				state_ERROR);
		{ok, "\"", Arg2} -> read(Fun, Arg2, F, R, $",
				action_ERROR,
				state_ERROR);
		{ok, "\r", Arg2} -> read(Fun, Arg2, F, R, $\r,
				action_NONE,
				state_CR);
		{ok, "\n", Arg2} -> read(Fun, Arg2, F, R, $\n,
				action_NONE,
				state_RECORD_END);
		{ok, [C], Arg2} -> read(Fun, Arg2, F, R, C,
				action_ERROR,
				state_ERROR);
		{eof, Arg2} -> read(Fun, Arg2, F, R, eof,
				action_NONE,
				state_RECORD_END)
	catch
		_:Why -> {ioerror, Why}
	end;

read(_Fun, Arg, _F, R, state_RECORD_END) ->
	{ok, lists:reverse(R), Arg}.

read(Fun, Arg, F, R, _C, action_NONE, State) ->
	read(Fun, Arg, F, R, State);
read(Fun, Arg, F, R, C, action_APPEND, State) ->
	read(Fun, Arg, [C|F], R, State);
read(Fun, Arg, F, R, _C, action_FLUSH, State) ->
	read(Fun, Arg, [], [lists:reverse(F)|R], State);
read(_Fun, Arg, F, R, C, action_ERROR, _State) ->
	case C of
		eof ->
			{formaterror, eof, lists:reverse(F), lists:reverse(R), Arg};
		Ch ->
			{formaterror, [Ch], lists:reverse(F), lists:reverse(R), Arg}
	end.


%% @spec file(IoDevice) => {ok, [Char], IoDevice}.
%%                         {eof, IoDevice}.
file(IoDevice) ->
	case io:get_chars(IoDevice, "", 1) of
		[Char] ->
			{ok, [Char], IoDevice};
		eof ->
			{eof, IoDevice}
	end.


%% @spec string(String) => {ok, [Char], RestOfString}.
%%                         {eof, []}
string([Char|Str]) -> {ok, [Char], Str};
string([]) -> {eof, []}.

File: Ruby/lib/csvparser.rb
# coding: utf-8
#
# Copyright 2012 Norio Agawa
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

class CsvParser

	def initialize(io)
		@io = io
	end

	def read
		record = nil
		field = ""
		state = @@RECORD_BEGIN
		until state == @@RECORD_END
			ch = @io.getc
			trans = state.call(ch)
			case trans[:action]
			when :APPEND
				field << ch
			when :FLUSH
				record = [] if record.nil?
				record << field
				field = ""
			when :ERROR
				raise CsvError, "failed to parse"
			end
			state = trans[:state]
		end
		record
	end

	def read_records
		ret = []
		until (rec = read).nil?
			ret << rec
		end
		ret
	end

	def each
		until (rec = read).nil?
			yield rec
		end
	end

	alias :each_records :each

	@@RECORD_BEGIN = lambda {|ch|
		case ch
		when ","	# COMMA
			{:action => :FLUSH,  :state => @@FIELD_BEGIN}
		when "\""	# DQUOTE
			{:action => :NONE,   :state => @@ESCAPED}
		when "\r"	# CR
			{:action => :FLUSH,  :state => @@CR}
		when "\n"	# LF
			{:action => :FLUSH,  :state => @@RECORD_END}
		when nil	# EOF
			{:action => :NONE,   :state => @@RECORD_END}
		else		# TEXTDATA
			{:action => :APPEND, :state => @@NONESCAPED}
		end
	}

	@@FIELD_BEGIN = lambda {|ch|
		case ch
		when ","	# COMMA
			{:action => :FLUSH,  :state => @@FIELD_BEGIN}
		when "\""	# DQUOTE
			{:action => :NONE,   :state => @@ESCAPED}
		when "\r"	# CR
			{:action => :FLUSH,  :state => @@CR}
		when "\n"	# LF
			{:action => :FLUSH,  :state => @@RECORD_END}
		when nil	# EOF
			{:action => :FLUSH,  :state => @@RECORD_END}
		else		# TEXTDATA
			{:action => :APPEND, :state => @@NONESCAPED}
		end
	}

	@@NONESCAPED = lambda {|ch|
		case ch
		when ","	# COMMA
			{:action => :FLUSH,  :state => @@FIELD_BEGIN}
		when "\""	# DQUOTE
			{:action => :APPEND, :state => @@NONESCAPED}
		when "\r"	# CR
			{:action => :FLUSH,  :state => @@CR}
		when "\n"	# LF
			{:action => :FLUSH,  :state => @@RECORD_END}
		when nil	# EOF
			{:action => :FLUSH,  :state => @@RECORD_END}
		else		# TEXTDATA
			{:action => :APPEND, :state => @@NONESCAPED}
		end
	}

	@@ESCAPED = lambda {|ch|
		case ch
		when ","	# COMMA
			{:action => :APPEND, :state => @@ESCAPED}
		when "\""	# DQUOTE
			{:action => :NONE,   :state => @@DQUOTE}
		when "\r"	# CR
			{:action => :APPEND, :state => @@ESCAPED}
		when "\n"	# LF
			{:action => :APPEND, :state => @@ESCAPED}
		when nil	# EOF
			{:action => :ERROR,  :state => nil}
		else		# TEXTDATA
			{:action => :APPEND, :state => @@ESCAPED}
		end
	}

	@@DQUOTE = lambda {|ch|
		case ch
		when ","	# COMMA
			{:action => :FLUSH,  :state => @@FIELD_BEGIN}
		when "\""	# DQUOTE
			{:action => :APPEND, :state => @@ESCAPED}
		when "\r"	# CR
			{:action => :FLUSH,  :state => @@CR}
		when "\n"	# LF
			{:action => :FLUSH,  :state => @@RECORD_END}
		when nil	# EOF
			{:action => :FLUSH,  :state => @@RECORD_END}
		else		# TEXTDATA
			{:action => :ERROR,  :state => nil}
		end
	}

	@@CR = lambda {|ch|
		case ch
		when ","	# COMMA
			{:action => :ERROR,  :state => nil}
		when "\""	# DQUOTE
			{:action => :ERROR,  :state => nil}
		when "\r"	# CR
			{:action => :NONE,   :state => @@CR}
		when "\n"	# LF
			{:action => :NONE,   :state => @@RECORD_END}
		when nil	# EOF
			{:action => :NONE,   :state => @@RECORD_END}
		else		# TEXTDATA
			{:action => :ERROR,  :state => nil}
		end
	}

	@@RECORD_END = lambda {|ch|
		nil
	}

end

class CsvError < StandardError
end
